### ISSUE TITLE
Test clean up does not work due to deferring functions not being called

### ISSUE THREAD

**Issue by shibumi** [in-toto/in-toto-golang#60]
*Please fill in the fields below to submit an issue or feature request.  The
more information that is provided, the better.*

**Description of issue or feature request**:
Right now, we use the `TestMain` function for setting up our temporary test directories, copying the `test/data` directory over and cleaning it up. The problem is, that the deferring functions are not being called, due to calling `os.Exit(m.Run())` in the end.
This can be fixed either via using `Go 1.15` in August and just calling `m.Run()` without `os.Exit()` or via creating setup and teardown methods, like this:
```go

func setup() string {
	testDir, err := ioutil.TempDir("", "testdir")
	if err != nil {
		panic("Cannot create temp test dir")
	}
	err = os.Chdir(testDir)
	if err != nil {
		fmt.Printf("Unable to change dir to %s: %s", testDir, err)
	}
        return testDir
}

func teardown(testdir string) {
	if err := os.RemoveAll(testDir); err != nil {
		fmt.Printf("Unable to remove directory %s: %s", testDir, err)
	}
}

func TestMain(m *testing.M) {
	testdir := setup()
	code := m.Run()
	teardown(testdir)
	os.Exit(code)
}
```

**Current behavior**:
`TestMain` does not clean up properly after calling the tests.

**Expected behavior**:
The temporary test directories should be deleted, after calling the tests.

**Comment by lukpueh** [comment]
Great find, @shibumi! I didn't know that os.Exit skipped defers.  If we use a vanilla teardown function call instead of defer, like you propose above, then we somehow need to make sure to teardown if we exit prematurely, e.g. to do an undeferred panic.

**Comment by shibumi** [comment]
@lukpueh exactly, that's why I actually prefer to wait until Go 1.15, because this bug will get fixed in Go 1.15:

* the corresponding issue on upstream: https://github.com/golang/go/issues/34129
* corresponding PR: https://go-review.googlesource.com/c/go/+/219639/
